src/operators/skipUntil.ts

```diff
@@ -19,8 +19,9 @@ class SkipUntilOperator<T, R> implements Operator<T, R> {
 class SkipUntilSubscriber<T> extends Subscriber<T> {
   private notificationSubscriber: NotificationSubscriber<any> = null;
 
-  constructor(destination: Subscriber<T>, private notifier: Observable<any>) {
-    super(destination);
+  constructor(public destination: Subscriber<T>,
+              private notifier: Observable<any>) {
+    super(null);
```

Comment:
Sure.

So, first of all I don't fully understand why some things were built as they are in Subscriber with regards to the shared subscriptions. Would be good if someone can clarify what was the intent.

If `super(destination)` is used, the secret sauce in the constructor will set `this._subscription`.
Later, in `Subscriber.proto.unsubscribe()` it does:
```js
if (this._isUnsubscribed) {
  return;
} else if (this._subscription) {
  this._isUnsubscribed = true;
} else {
  super.unsubscribe();
}
```

`this._isUnsubscribed` is false, so we go to the second if check, and `this._subscription` is truthy, so we just end up setting `this._isUnsubscribed` to true, but not actually unsubscribing the current Subscriber from the Observable.

So this test would not pass:
```js
  it('should skip all element and does not complete when another observable never completes', function () {
    var e1 =   hot( '--a--b--c--d--e--|');
    var e1subs =    '^                !'; // <== unsubscribe should happen, but never does!
    var skip = cold('-');
    var skipSubs =  '^                !';
    var expected =  '-';

    expectObservable(e1.skipUntil(skip)).toBe(expected);
    expectSubscriptions(e1.subscriptions).toBe(e1subs);
    expectSubscriptions(skip.subscriptions).toBe(skipSubs);
  });
```

The inner subscription for the `skip` Observable is correctly unsubscribed when `e1` completes, but the subscription to `e1` is **not** unsubscribed when `e1` completes. Because of how `Subscriber.proto.unsubscribe()` was coded as I explained above.

PS: you already merged some PRs where I used the same technique: https://github.com/ReactiveX/RxJS/pull/562/files#diff-0864900d001ae5c1b7070f19e5e13a2eR33.